### ISSUE TITLE
Stop referring to ungrouped columns from outer query in subquery

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#27266]
PostgreSQL seems to be particularly strict around subqueries and GROUP BY. The following query (full repro below):

```c#
from s in (from i in ctx.Invoices
        group i by i.History.Month
        into g
        select new
        {
            Month = g.Key,
            Total = g.Sum(p => p.Amount),
        })
    select new
    {
        s.Month,
        s.Total,
        Payment = ctx.Payments.Where(p => p.History.Month == s.Month).Sum(p => p.Amount)
    }
```

... generates the following SQL on PostgreSQL:

```sql
SELECT date_part('month', i."History")::INT AS "Month", COALESCE(SUM(i."Amount"), 0)::INT AS "Total", (
    SELECT COALESCE(SUM(p."Amount"), 0)::INT
    FROM "Payments" AS p
    WHERE date_part('month', p."History")::INT = date_part('month', i."History")::INT) AS "Payment"
FROM "Invoices" AS i
GROUP BY date_part('month', i."History")::INT
```

... and errors with the following: `subquery uses ungrouped column "i.History" from outer query`.

The same works fine on SQL Server, SQLite and MariaDB.

Note also that the current SQL is less efficient even when supported, since the same expression gets evaluated more than once (see https://github.com/dotnet/efcore/issues/27266#issuecomment-1022365293).

See suggested alternative SQL by @mojtabakaviani in https://github.com/npgsql/efcore.pg/issues/2236#issue-1112392955.

<details>
<summary>Full repro</summary>

```c#
await using var ctx = new BlogContext();
await ctx.Database.EnsureDeletedAsync();
await ctx.Database.EnsureCreatedAsync();

var query = from s in (from i in ctx.Invoices
        group i by i.History.Month
        into g
        select new
        {
            Month = g.Key,
            Total = g.Sum(p => p.Amount),
        })
    select new
    {
        s.Month,
        s.Total,
        Payment = ctx.Payments.Where(p => p.History.Month == s.Month).Sum(p => p.Amount)
    };

_ = query.ToList();

public class BlogContext : DbContext
{
    public DbSet<Invoice> Invoices { get; set; }
    public DbSet<Payment> Payments { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            // .UseSqlServer(@"Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0")
            .UseNpgsql(@"Host=localhost;Username=test;Password=test")
            // .UseSqlite("Data Source=/tmp/foo.sql")
            // .UseMySql("Server=localhost;user=test;password=test;database=test", new MariaDbServerVersion("10.5"))
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();
}

public class Invoice
{
    public int Id { get; set; }
    public int Amount { get; set; }

    public DateTime History { get; set; }
}

public class Payment
{
    public int Id { get; set; }
    public int Amount { get; set; }

    public DateTime History { get; set; }
}
```

</details>

Raised by @mojtabakaviani in https://github.com/npgsql/efcore.pg/issues/2236

**Comment by smitpatel** [comment]
Cannot really do much, probably something which is just unsupported.

**Comment by mojtabakaviani** [comment]
Yes, but if EF Core avoid convert query to join work properly.

**Comment by smitpatel** [comment]
There is no conversion to join here.

**Comment by mojtabakaviani** [comment]
@smitpatel  I mean, converting nested query to scalar result, for this example.

**Comment by smitpatel** [comment]
But there is no conversion EF Core performs here. It is nested query which happens to be scalar result. Every scalar results are same regardless of where they come from.

**Comment by roji** [comment]
Note @mojtabakaviani alternate SQL proposal in https://github.com/npgsql/efcore.pg/issues/2236:

```sql
SELECT i0."Month", i0."Total", (
    SELECT COALESCE(SUM(p.amount), 0.0) FROM payments AS p
    WHERE date_part('month', p.history)::INT = i0."Month"::INT) AS "Payment"
FROM (
    SELECT date_part('month', i.history)::INT AS "Month", COALESCE(SUM(i.amount), 0.0) AS "Total"
    FROM invoices AS i
    GROUP BY date_part('month', i.history)::INT) AS i0;
```

It's also possible to extract the month out with a WITH as follows:

```c#
WITH "InvoiceData" AS (
    SELECT date_part('month', i."History")::INT AS "Month", i."Amount"
    FROM "Invoices" AS i
)
SELECT id."Month", COALESCE(SUM(id."Amount"), 0)::INT AS "Total", (
    SELECT COALESCE(SUM(p."Amount"), 0)::INT
    FROM "Payments" AS p
    WHERE date_part('month', p."History")::INT = id."Month") AS "Payment"
FROM "InvoiceData" AS id
GROUP BY "Month";
```

Both these solutions should also be better from a perf standpoint (also for SQL Server), as the month is calculated only once rather than three times in the current query. So it would be good if we could do something here.

**Comment by mojtabakaviani** [comment]
Thanks @roji for proposal, I test and work in the both SQL Server and Postgresql. also solve nested query issue #27267

**Comment by roji** [comment]
@mojtabakaviani I ended up removing the WITH alternative since it doesn't really add anything that a simple subquery doesn't already accomplish (as in your original proposal).

**Comment by ajcvickers** [comment]
Note from triage: add tests for this case.

**Comment by roji** [comment]
@ajcvickers note that this has what could be an important perf aspect as well (multiple function/subquery evaluation which could be avoided), so this isn't just about finding a solution for PG. We could add a test here, and have another issue tracking changing the SQL we generate (which would both help perf and make the query work on PG).

**Comment by roji** [comment]
Great! In that case I guess this can just remind us to add the test.

**Comment by smitpatel** [comment]
Poaching this. Pending selector work is on hold now. https://github.com/dotnet/efcore/issues/27433 has same issue for SqlServer in different scenario. Also #27132 work detected that we may not be able to lift always causing subquery. An ideal fix here would be to basically pushdown into subquery when grouping key is complex to avoid issues generated by subquery having correlation predicate with grouping key. Does that sound good to you @roji ?
